File: src/example/src/serial_write.cpp
#include <iostream>
#include <string>
#include <boost/asio.hpp>

using namespace std;

string speed(float f_velocity) {
    stringstream strs;
    char buff[100];
    snprintf(buff, sizeof(buff), "%.2f;;\r\n", f_velocity);
    strs << "#" << "1" << ":" << buff;
    return strs.str();
}

string steer(float f_angle) {
    stringstream strs;
    char buff[100];
    snprintf(buff, sizeof(buff), "%.2f;;\r\n", f_angle);
    strs << "#" << "2" << ":" << buff;
    return strs.str();
}

int main() {
    boost::asio::io_service io;
    boost::asio::serial_port serial(io);

    serial.open("/dev/ttyACM0");
    serial.set_option(boost::asio::serial_port_base::baud_rate(19200));

    string speed_msg = speed(0.0);
    cout << "speed msg sent: " << speed_msg << endl;
    boost::asio::write(serial, boost::asio::buffer(speed_msg));

    string steer_msg = steer(0.0);
    cout << "steer msg sent: " << steer_msg << endl;
    boost::asio::write(serial, boost::asio::buffer(steer_msg));

    return 0;
}

File: src/example/src/serial_write.py
import serial

def speed(f_velocity):
    return "#{0}:{1:.2f};;\r\n".format(1, f_velocity)

def steer(f_angle):
    return "#{0}:{1:.2f};;\r\n".format(2, f_angle)

def both(f_velocity,f_angle):
    return "#{0}:{1:.2f}:{1:.2f};;\r\n".format(8, f_velocity, f_angle)

def main():
    serial_port = serial.Serial('/dev/ttyACM0', baudrate=115200)

    speed_msg = speed(10.0)
    print("speed msg sent:", speed_msg)
    serial_port.write(speed_msg.encode())

    steer_msg = steer(10.0)
    print("steer msg sent:", steer_msg)
    serial_port.write(steer_msg.encode())

    both_msg = both(10.0, 10.0)
    print("both msg sent:", both_msg)
    serial_port.write(both_msg.encode())

    serial_port.close()

if __name__ == "__main__":
    main()
